text/0000-computed-context.md

```diff
@@ -0,0 +1,329 @@
+
+Computed values _cannot_ be assigned to; they are read-only values.
+
+Computed values _extend_ `context` values, so their keys cannot be the same as any `context` keys.
+
+Just like getters, computed values are computed on _read_. Caching optimizations may be possible in the future: if we are able to determine which context dependencies a computed value has, we can invalidate the cached computed context value whenever any of those dependencies change on `assign(...)`.
+
+**Specifying computed context**
+
+```ts
+const machine = createMachine({
+  context: {
+    firstName: '',
+    lastName: ''
+  },
+  computed: {
+    // Computed from non-computed context
+    fullName: ({ context }) => `${context.firstName} ${context.lastName}`,
+    // Computed from computed context
+    isLongName: ({ context }) => context.fullName.length > 20,
+    // Computed from state and context
+    errorMessage: (state) => {
+      if (state.hasTag('error')) {
+        return state.context.fullName.trim().length === 0
+          ? 'Invalid name'
+          : 'Unknown error'
+      }
+
+      return null;
+    },
+    // Returns a function, e.g.:
+    // `const greeting = state.context.greet('Hello')`
+    greet: ({ context }) => (greeting: string) => `${greeting}, ${context.fullName}`
+  }
+});
+```
+
+**Reading computed context**
+
+```js
+// ...
+
+const name = state.context.fullName; // computed from context.firstName and context.lastName
+
+const greeting = state.context.greet('Hello');
+```
+
+**Usage in assign**
+
+```ts
+const machine = createMachine({
+  context: {
+    firstName: '',
+    lastName: '',
+    age: 0
+  },
+  computed: {
+    fullName: ({ context }) => `${context.firstName} ${context.lastName}`,
+    // ...
+  },
+  // ...
+  entry: assign({
+    firstName: (_, event) => event.value,
+    age: (context, event) => {
+      // This will be the full name *prior* to the assignment above
+      // since it is within the same assign() call
+      console.log(context.fullName);
+
+      return event.value;
+    }
+  }),
+  // ...
+  exit: [
+    assign({ firstName: (_, event) => event.value }),
+    assign({
+      age: (context, event) => {
+        // This will be the full name *after* the previous assignment
+        // since it is a subsequent assign() call
+        console.log(context.fullName);
+
+        return event.value;
+      }
+    })
+  ],
+  // ...
+  on: {
+    someEvent: {
+      actions: assign({
+        // ❌ This will be a type error
+        fullName: 'something else'
+      })
+    }
+  }
+});
+```
+
+## How we teach this
+
+- We add a new "Computed Context" section to the existing guide on machine context
+- Computed context is introduced as "read-only context" where the computed values are determined from a function that takes in the `state` and returns the value.
+- Teach that computed context should not be assigned to
```

Comment:
What would technically happen if you try to assign to a computed value? We can mention them here to let the reader know what to expect.